### ISSUE TITLE
Build error with glibc 2.35

### ISSUE THREAD

**Issue by cpaelzer** [vmware/open-vm-tools#570]
### Describe the bug

FTBFS when building from source in Ubuntu:

````
libtool: compile:  gcc -DPACKAGE_NAME=\"open-vm-tools\" -DPACKAGE_TARNAME=\"open-vm-tools\" -DPACKAGE_VERSION=\"11.3.5\" "-DPACKAGE_STRING=\"open-vm-tools 11.3.5\"" -DPACKAGE_BUGREPORT=\"open-vm-tools-devel@lists.sourceforge.net\" -DPACKAGE_URL=\"\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_TIME_H=1 -DSTDC_HEADERS=1 -DPACKAGE=\"open-vm-tools\" -DVERSION=\"11.3.5\" -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_X11_SM_SMLIB_H=1 -DHAVE_X11_ICE_ICELIB_H=1 -DHAVE_X11_EXTENSIONS_XCOMPOSITE_H=1 -DHAVE_DLOPEN=1 -DHAVE_ECVT=1 -DHAVE_FCVT=1 -DNO_DNET=1 -DHAVE_CRYPT_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_WCHAR_H=1 -DHAVE_SYS_IO_H=1 -DHAVE_SYS_PARAM_H=1 -DHAVE_SYS_SYSINFO_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_USER_H=1 -DHAVE_SYS_VFS_H=1 -DHAVE_UNWIND_H=1 -DHAVE__BOOL=1 -DHAVE_STDBOOL_H=1 -DHAVE_STRUCT_STAT_ST_RDEV=1 -DTIME_WITH_SYS_TIME=1 -I. -I/root/open-vm-tools/open-vm-tools/lib/include -I/root/open-vm-tools/open-vm-tools/lib/include -DUSING_AUTOCONF=1 -DOPEN_VM_TOOLS -DUSE_VGAUTH -DNO_ICU -DVMX86_TOOLS -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_XOPEN_SOURCE=500 -D_BSD_SOURCE -D_SVID_SOURCE -D_DEFAULT_SOURCE -DENABLE_RESOLUTIONKMS -DUSE_SSL_DIRECT -fPIC -Wno-error=deprecated-declarations -Wno-error=address-of-packed-member -Wno-nonnull -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wall -Werror -Wno-pointer-sign -Wno-unused-value -fno-strict-aliasing -Wno-unknown-pragmas -Wno-uninitialized -Wno-deprecated-declarations -Wno-unused-const-variable -Wno-unused-but-set-variable -MT asyncsocket.lo -MD -MP -MF .deps/asyncsocket.Tpo -c asyncsocket.c  -fPIC -DPIC -o .libs/asyncsocket.o
In file included from /usr/include/features.h:486,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdio.h:27,
                 from asyncsocket.c:45:
In function ‘poll’,
    inlined from ‘AsyncTCPSocketPollWork’ at asyncsocket.c:2861:19:
/usr/include/x86_64-linux-gnu/bits/poll2.h:39:10: error: ‘__poll_chk’ specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
   39 |   return __glibc_fortify (poll, __nfds, sizeof (*__fds),
      |          ^~~~~~~~~~~~~~~
In file included from /usr/include/x86_64-linux-gnu/sys/poll.h:87,
                 from asyncsocket.c:74:
/usr/include/x86_64-linux-gnu/bits/poll2.h: In function ‘AsyncTCPSocketPollWork’:
/usr/include/x86_64-linux-gnu/bits/poll2.h:28:12: note: in a call to function ‘__poll_chk’ declared with attribute ‘access (write_only, 1, 2)’
   28 | extern int __poll_chk (struct pollfd *__fds, nfds_t __nfds, int __timeout,
      |            ^~~~~~~~~~
cc1: all warnings being treated as errors
```

That only happens when a few conditions are met to make the compiler spot that issue at build time. That is LTO (to find the conflict between libc6 and open-vm-tools) optimization (to pre-cacle the sizes) and a new glibc 2.35.

### Reproduction steps

```bash
1. get glibc 2.35 e.g. in Ubuntu 22.04 -proposed right now there is  2.35-0ubuntu1 
2. get open-vm-tools from git
3. Build it with
 * LTO: -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects
 * Optimizing with gcc: -O2
 * gcc 11
```


### Expected behavior

I'd like to ask you to analyze the case and get the open-vm-tools code fixed to not trigger that issue.

For the time until a proper fix the following is a valid workaround silencing the warning for just this place in the code.

```
diff --git a/open-vm-tools/lib/asyncsocket/asyncsocket.c b/open-vm-tools/lib/asyncsocket/asyncsocket.c
index 05147d2e..546932ec 100644
--- a/open-vm-tools/lib/asyncsocket/asyncsocket.c
+++ b/open-vm-tools/lib/asyncsocket/asyncsocket.c
@@ -2850,6 +2850,10 @@ AsyncTCPSocketPollWork(AsyncTCPSocket **asock,     // IN:
          pfd[i].events = read ? POLLIN : POLLOUT;
       }
 
+#if defined(__GNUC__) && (__GNUC__ >= 11)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-overflow"
+#endif
       if (parentSock != NULL) {
          AsyncTCPSocketUnlock(parentSock);
          retval = poll(pfd, numSock, timeoutMS);
@@ -2863,6 +2867,9 @@ AsyncTCPSocketPollWork(AsyncTCPSocket **asock,     // IN:
             AsyncTCPSocketLock(asock[i]);
          }
       }
+#if defined(__GNUC__) && (__GNUC__ >= 11)
+#pragma GCC diagnostic pop
+#endif
 #else
       tv.tv_sec = timeoutMS / 1000;
       tv.tv_usec = (timeoutMS % 1000) * 1000;
```

### Additional context

_No response_

**Comment by cpaelzer** [comment]
Please note that you use `int numSock` that you pass into a [nfds_t](https://pubs.opengroup.org/onlinepubs/007904975/basedefs/poll.h.html) which is [supposed to be unsigned int](https://pubs.opengroup.org/onlinepubs/007904975/basedefs/poll.h.html).

That seems suspicious to and might be be related.

**Comment by jonathanvmw** [comment]
@cpaelzer 

Thank you for reporting the issue.   Coincidentally, we already had an internal bug report for this and have been working on a fix.

**Comment by johnwvmw** [comment]
A fix for this compilation warning issue has been checked into the "devel" branch at https://github.com/vmware/open-vm-tools/commit/de6d129476724668b8903e2a87654f50ba21b1b2.

That fix can be used with open-vm-tools 11.3.0, 11.3.5 and the upcoming 12.0.0 source bundles.   Note, the fix was not in time to make it in the 12.0.0 open-vm-tools release; it will be incorporated into the fixes for the future open-vm-tools 12.0.5 release.

**Comment by johnwvmw** [comment]
As indicated in the previous comment, the fix is in the development branch and can be directly applied to the open-vm-source for 11.3.0, 11.3.5 and 12.0.0.   The open-vm-tools 12.0.0 Release Notes will highlight that this may be encountered when building the 12.0.0 open-vm-tools and the fix https://github.com/vmware/open-vm-tools/commit/de6d129476724668b8903e2a87654f50ba21b1b2 can be used.

Closing this issue - fixed.
